lib/msal-node/src/client/ClientApplication.ts

```diff
@@ -279,33 +282,132 @@ export abstract class ClientApplication {
+                    if (persistence) {
+                        this.logger.info(
+                            "Overwriting in-memory cache with persistent cache"
+                        );
+                        this.storage.clear();
+                        const cacheContext = new TokenCacheContext(
+                            this.tokenCache,
+                            false
+                        );
+                        await persistence.beforeCacheAccess(cacheContext);
+                        const cacheSnapshot =
+                            this.tokenCache.getCacheSnapshot();
+                        this.storage.setCache(cacheSnapshot);
+                        await persistence.afterCacheAccess(cacheContext);
+
+                        this.logger.info("Searching again for a valid token");
+                        return await this.acquireCachedToken(
+                            validRequest,
+                            silentFlowClient,
+                            clientConfiguration
+                        );
+                    }
+                } catch {
+                    if (
+                        e instanceof ClientAuthError &&
+                        e.errorCode ===
+                            ClientAuthErrorCodes.tokenRefreshRequired
+                    ) {
+                        const refreshTokenClient = new RefreshTokenClient(
+                            clientConfiguration
+                        );
+                        return refreshTokenClient.acquireTokenByRefreshToken(
+                            validRequest
+                        );
+                    } else if (e instanceof AuthError) {
+                        e.setCorrelationId(validRequest.correlationId);
+                    }
+                    serverTelemetryManager.cacheFailedRequest(e as AuthError);
+                    throw e;
+                }
+
+                if (
+                    e instanceof ClientAuthError &&
+                    e.errorCode === ClientAuthErrorCodes.tokenRefreshRequired
+                ) {
+                    const refreshTokenClient = new RefreshTokenClient(
+                        clientConfiguration
+                    );
+                    return refreshTokenClient.acquireTokenByRefreshToken(
+                        validRequest
+                    );
+                } else if (e instanceof AuthError) {
+                    e.setCorrelationId(validRequest.correlationId);
+                }
+                serverTelemetryManager.cacheFailedRequest(e as AuthError);
+                throw e;
+            }
         } catch (e) {
             if (e instanceof AuthError) {
                 e.setCorrelationId(validRequest.correlationId);
             }
-            serverTelemetryManager.cacheFailedRequest(e as AuthError);
+            serverTelemetryManager.cacheFailedRequest(e);
             throw e;
         }
     }
 
+    private async acquireCachedToken(
```

Comment:
Should this be renamed to acquireCachedTokenSilent to indicate that it's only being used by silent flow?